common/src/Renderer/EntityDecalRenderer.cpp

```diff
@@ -0,0 +1,438 @@
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ TrenchBroom is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with TrenchBroom. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "EntityDecalRenderer.h"
+
+#include "Assets/DecalDefinition.h"
+#include "Assets/Texture.h"
+#include "Assets/TextureManager.h"
+#include "BrushRendererArrays.h"
+#include "Model/BrushFace.h"
+#include "Model/BrushNode.h"
+#include "Model/EditorContext.h"
+#include "Model/EntityNode.h"
+#include "Model/ModelUtils.h"
+#include "Model/Polyhedron_Face.h"
+#include "Model/TexCoordSystem.h"
+#include "Model/WorldNode.h"
+#include "View/MapDocument.h"
+#include "octree.h"
+
+#include <kdl/memory_utils.h>
+#include <kdl/overload.h>
+
+#include <vecmath/intersection.h>
+
+#include <cstring>
+
+namespace TrenchBroom
+{
+namespace Renderer
+{
+EntityDecalRenderer::EntityDecalRenderer(const std::weak_ptr<View::MapDocument>& document)
+  : m_document{document}
+{
+  clear();
+}
+
+void EntityDecalRenderer::invalidate()
+{
+  for (auto& [ent, data] : m_entities)
+  {
+    invalidateDecalData(data);
+  }
+}
+
+void EntityDecalRenderer::clear()
+{
+  m_entities.clear();
+  m_vertexArray = std::make_shared<BrushVertexArray>();
+  m_faces = std::make_shared<TextureToBrushIndicesMap>();
+  m_faceRenderer = FaceRenderer{m_vertexArray, m_faces, m_faceColor};
+}
+
+void EntityDecalRenderer::updateNode(Model::Node* node)
+{
+  node->accept(kdl::overload(
+    [](Model::WorldNode*) {},
+    [](Model::LayerNode*) {},
+    [](Model::GroupNode*) {},
+    [&](const Model::EntityNode* entity) { updateEntity(entity); },
+    [&](const Model::BrushNode* brush) { updateBrush(brush); },
+    [](Model::PatchNode*) {}));
+}
+
+void EntityDecalRenderer::removeNode(Model::Node* node)
+{
+  node->accept(kdl::overload(
+    [](Model::WorldNode*) {},
+    [](Model::LayerNode*) {},
+    [](Model::GroupNode*) {},
+    [&](const Model::EntityNode* entity) { removeEntity(entity); },
+    [&](const Model::BrushNode* brush) { removeBrush(brush); },
+    [](Model::PatchNode*) {}));
+}
+
+std::optional<Assets::DecalSpecification> EntityDecalRenderer::getDecalSpecification(
+  const Model::EntityNode* entityNode)
+{
+  const auto decalSpec = entityNode->entity().decalSpecification();
+  return decalSpec.textureName.empty() ? std::nullopt : std::make_optional(decalSpec);
+}
+
+void EntityDecalRenderer::updateEntity(const Model::EntityNode* entityNode)
+{
+  // if the entity isn't visible, don't create decal geometry for it
+  const auto& editorContext = kdl::mem_lock(m_document)->editorContext();
+  if (!editorContext.visible(entityNode))
```

Comment:
Yep you're right - I had this check in the `getDecalSpecification` method before but moved it into here so I could keep that method `static` (and because I felt it made more sense for this method to always return the entity's decal specification regardless of its visibility state). After changing that I forgot to re-test visibility stuff!

I think my fix here would be:

```cpp
  // if the entity isn't visible, don't create decal geometry for it
  const auto& editorContext = kdl::mem_lock(m_document)->editorContext();

  // check if the entity has a decal specification
  const auto spec =
    editorContext.visible(entityNode) ? getDecalSpecification(entityNode) : std::nullopt;
```